### PR TITLE
JAMES-3775 Write a webadmin route to feed spam to RSpamD

### PR DESCRIPTION
**Pull request by quantranhong1999**
TODO task unit tests, webadmin route , webadmin unit tests, and a bit of documentation the route.

### CODE REVIEW

**Review by Arsnael — COMMENTED** [review]
Read it

**Review by Arsnael — APPROVED** [review]

File: third-party/rspamd/README.md
Comment (chibenwa) [inline third-party/rspamd/README.md:14]: Please allow to provide unit. EG:

`curl -XPOST 'http://ip:port/rspamd?action=reportSpam&period=1d`

This makes it way more friendly

File: third-party/rspamd/README.md
Comment (chibenwa) [inline third-party/rspamd/README.md:23]: Link to https://james.apache.org/server/manage-webadmin.html#Endpoints_returning_a_task instead (as endpoints returning tasks is something generic, you do not need to detail it here).

File: third-party/rspamd/README.md
Comment (chibenwa) [inline third-party/rspamd/README.md:40]: What are possible response codes?

File: third-party/rspamd/pom.xml
Comment (chibenwa) [inline third-party/rspamd/pom.xml:87]: Quick question: why is this dependency needed?

File: third-party/rspamd/src/main/java/org/apache/james/rspamd/route/FeedMessageRoute.java
Comment (chibenwa) [inline third-party/rspamd/src/main/java/org/apache/james/rspamd/route/FeedMessageRoute.java:96]: Use DurationParser?

File: third-party/rspamd/src/main/java/org/apache/james/rspamd/task/FeedSpamToRSpamDTask.java
Comment (chibenwa) [inline third-party/rspamd/src/main/java/org/apache/james/rspamd/task/FeedSpamToRSpamDTask.java:388]: if we have a probability of `1.0` we probably don't need a call to random.

File: third-party/rspamd/src/main/java/org/apache/james/rspamd/task/FeedSpamToRSpamDTask.java
Comment (chibenwa) [inline third-party/rspamd/src/main/java/org/apache/james/rspamd/task/FeedSpamToRSpamDTask.java:352]:
```suggestion
                    .forOperation(messageResult1 -> messageResult -> rSpamDHttpClient.reportAsSpam(messageResult.getHeaders().getInputStream(), messageResult.getBody().getInputStream())))
```

Please simpify this.

File: third-party/rspamd/src/main/java/org/apache/james/rspamd/task/GetMailboxMessagesService.java
Comment (chibenwa) [inline third-party/rspamd/src/main/java/org/apache/james/rspamd/task/GetMailboxMessagesService.java:55]: The use of FULL_CONTENT here will load all the messages content, all the time.

This means that this task will always load ALL the content of SPAM mailbox, which can be avoided with sampling/period parameters.

I suggest only loading metadata here and loading content upon reporting ham/spam.

File: third-party/rspamd/pom.xml
Comment (quantranhong1999) [reply third-party/rspamd/pom.xml:87]: `RSpamDModule` to specify webadmin route bean and DTO modules (just not want one forget which to add).

Comment (chibenwa) [reply third-party/rspamd/pom.xml:87]: Guice only is not enough?

File: third-party/rspamd/src/main/java/org/apache/james/rspamd/task/FeedSpamToRSpamDTask.java
Comment (quantranhong1999) [reply third-party/rspamd/src/main/java/org/apache/james/rspamd/task/FeedSpamToRSpamDTask.java:352]: Switching to this, we would have to do a verbose try catch to use `.fullContent`. I put in some effort but did not find a better solution.

Comment (chibenwa) [reply third-party/rspamd/src/main/java/org/apache/james/rspamd/task/FeedSpamToRSpamDTask.java:352]: Yes but we could wrap this using `Throwing.function` in the right place, right?

**Comment by chibenwa** [comment]
It would have been easier for me to follow if you cherry-picked #1100 then have a fix up to adapt your PR to it.

With a force push I need to re-read everything yet-again, and I will not do it today (huge work, and it is late). 

This force push takes up my time and creates (unecessary) delays on your work, and a rebase was needed anyway to get rid of the no-review commit.

**Comment by quantranhong1999** [comment]
You just need to read the latest commit (adapt Benoit review), the rest is the same.

**Review by chibenwa — APPROVED** [review]
Looks good to me!

**Comment by chibenwa** [comment]
Yes I missed the "adapt benoit review". Sorry I may still be a bit jet-lagged.

File: third-party/rspamd/src/main/java/org/apache/james/rspamd/route/FeedMessageRoute.java
Comment (vttranlina) [inline third-party/rspamd/src/main/java/org/apache/james/rspamd/route/FeedMessageRoute.java:64]: I don't think we already have a singleton for `MessageMapper`, 
MessageMapper is created for each MailboxSession. 

-> 
```
MailboxSessionMapperFactory.getMessageMapper(MailboxSession)
```

Comment (quantranhong1999) [reply third-party/rspamd/src/main/java/org/apache/james/rspamd/route/FeedMessageRoute.java:64]: Good remark. Thanks!

Comment (quantranhong1999) [reply third-party/rspamd/src/main/java/org/apache/james/rspamd/route/FeedMessageRoute.java]: https://github.com/apache/james-project/pull/1092/commits/05257991d344401e8aca278e8d2637a692705f4b solves this.

**Comment by Arsnael** [comment]
I don't think that Adapt Benoit review and Adapt Tung review are proper commit names though... :P

**Comment by quantranhong1999** [comment]
> I don't think that Adapt Benoit review and Adapt Tung review are proper commit names though... :P

Otherwise use fixup it is the same anyway. Yeh I just wait for the approval and rebase.

**Comment by Arsnael** [comment]
I disagree... fixups are a standard and easy to spot when you squash up. This... So ok it's a small PR but on a bigger scale can easily be missed :)

**Review by Arsnael — APPROVED** [review]

File: third-party/rspamd/pom.xml
Comment (vttranlina) [inline third-party/rspamd/pom.xml:132]: -> `<groupId>${james.groupId}</groupId>`

**Review by vttranlina — APPROVED** [review]

**Pull request merged**
